ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpResponse.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.statement
+
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.util.*
+import io.ktor.util.date.*
+import io.ktor.utils.io.*
+import kotlinx.coroutines.*
+
+/**
+ * A response for [HttpClient], second part of [HttpClientCall].
+ */
+abstract class HttpResponse : HttpMessage, CoroutineScope {
```

Comment:
Why not an interface?

Reply:
It could be. We must consider use cases.

Reply:
We need to think about it since "abstract class <-> interface" conversion is not a binary compatible change. The other think we need to notice is that we can restrict users from inheriting an abstract class while we can't do the same for an interface.

The first question is why one would like to inherit it?

Reply:
I would like to keep it as an abstract class here: 
  - there are no use cases to use it with other abstract classes 
  - we have more flexibility in future wo breaking compatibility

Reply:
Shouldn't we restrict inheritance for now? (e.g. by making the constructor internal or something)